### PR TITLE
Migrate device controller python binding to new controller and channel API

### PR DESCRIPTION
**Pull request by kghost**
#### Problem

Migrate to messaging layer #4170

 #### Summary of Changes

Same purpose as **Migrate device controller to messaging layer API** #4171, but migrating chip-tool will break lots of CI test, so instead of migrating chip-tool, we can start from the python binding of device controller.

I'm going to migrate the python binding to the new controller class, instead of the deprecated class.

### CODE REVIEW

**Review by rwalker-apple — APPROVED** [review]

File: src/messaging/Channel.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Comment (jelderton) [inline src/messaging/Channel.cpp:3]: nit: 2021

File: src/messaging/Channel.h [hunk 2]
```diff
@@ -0,0 +1,247 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Comment (jelderton) [inline src/messaging/Channel.h:3]: nit: 2021

File: src/messaging/ChannelContext.cpp [hunk 3]
```diff
@@ -0,0 +1,407 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Comment (jelderton) [inline src/messaging/ChannelContext.cpp:3]: nit: 2021

File: src/messaging/ChannelContext.h [hunk 4]
```diff
@@ -0,0 +1,154 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
```
Comment (jelderton) [inline src/messaging/ChannelContext.h:3]: nit: 2021

**Review by jelderton — APPROVED** [review]

**Comment by woody-apple** [comment]
@kghost can you take a peek at merge conflicts?

**Comment by woody-apple** [comment]
@pan-apple Can you take a peek at this? These seem like changes beyond just a python binding.

File: src/controller/python/chip-device-ctrl.py [hunk 5]
```diff
@@ -101,7 +101,7 @@ def __init__(self, rendezvousAddr=None):
 
         self.bleMgr = None
 
-        self.devCtrl = ChipDeviceCtrl.ChipDeviceController()
+        self.devCtrl = ChipDeviceCtrl.ChipDeviceController(112233)
```
Comment (woody-apple) [inline src/controller/python/chip-device-ctrl.py:104]: Hardcoded number?

**Review by woody-apple — CHANGES_REQUESTED** [review]
@kghost can you sync up with @vivien-apple about how to code-gen the python platform from ZAP? Lots of this appears hand rolled, which I don't think is something we want to do.

**Comment by kghost** [comment]
This PR includes #4019, the second commit is python binding part.

**Review by pan-apple — CHANGES_REQUESTED** [review]
@kghost , this PR is also bringing in Channel API, which seems unrelated to the purpose of this PR. Suggest using a separate PR for reviewing the Channel API.

File: src/controller/python/chip/Device.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+from abc import ABC, abstractmethod
+from ctypes import *
+from enum import Enum
+
+class DeviceStatusDelegate(ABC):
+    OnStatusChangeFunction = CFUNCTYPE(None)
+
+    def __init__(self, dmLib):
+        self.__dmLib = dmLib
```
Comment (mspang) [inline src/controller/python/chip/Device.py:31]: What is dmLib?

(please rename this)

File: src/controller/python/chip/Device.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+        def HandleStatusChange():
+            self.OnStatusChange()
+
+        self.__delegate = c_void_p(None)
+        # hold the reference to the callback to prevent it from being disposed.
+        self.handleOnStatusChange = DeviceStatusDelegate.OnStatusChangeFunction(HandleStatusChange)
+        self.__dmLib.nl_Chip_DeviceStatusDelegate_Create(pointer(self.__delegate), self.handleOnStatusChange);
+        pass
```
Comment (mspang) [inline src/controller/python/chip/Device.py:40]: Pass not needed here.

**Pull request closed without merging**